python/GafferSceneUI/TweakPlugValueWidget.py

```diff
@@ -0,0 +1,215 @@
+#
+##########################################################################
+
+import functools
+import imath
+
+import IECore
+import IECoreScene
+
+import Gaffer
+import GafferUI
+import GafferScene
+import GafferSceneUI
+import weakref
+
+
+class TweakPlugValueWidget( GafferUI.PlugValueWidget ) :
+
+	def __init__( self, plug ) :
+
+		# TODO - would be nice if this stuff didn't need to be added to the instance as metadata
+		# John has said that in the future we may be able to use match patterns to register metadata
+		# to children of a plug type, like we can do with nodes.
+
+		Gaffer.Metadata.registerValue( plug['name'], "description",
+			"The name of the parameter to apply the tweak to.", persistent=False
+		)
+		Gaffer.Metadata.registerValue( plug['mode'], "plugValueWidget:type",
+			"GafferUI.PresetsPlugValueWidget", persistent=False
+		)
+
+
+		# TODO - these presets should be set based on the type of the value plug - it doesn't make
+		# sense to offer the "Multiply" option on strings.  We could use dynamic menus if we sorted
+		# out the issue above.
+		for name, value in GafferScene.TweakPlug.Mode.names.items():
+			Gaffer.Metadata.registerValue( plug['mode'], "preset:" + name, value, persistent = False )
+
+		self.__row = GafferUI.ListContainer( GafferUI.ListContainer.Orientation.Horizontal, spacing = 4 )
+
+		GafferUI.PlugValueWidget.__init__( self, self.__row, plug )
+
+
+		nameWidget = GafferUI.StringPlugValueWidget( plug["name"] )
+		nameWidget.textWidget()._qtWidget().setFixedWidth( GafferUI.PlugWidget.labelWidth() )
+
+		self.__row.append( nameWidget,
+			verticalAlignment = GafferUI.Label.VerticalAlignment.Top
+		)
+
+		self.__row.append(
+			GafferUI.BoolPlugValueWidget(
+				plug["enabled"],
+				displayMode = GafferUI.BoolWidget.DisplayMode.Switch
+			),
+			verticalAlignment = GafferUI.Label.VerticalAlignment.Top,
+		)
+
+		self.__row.append( GafferUI.PlugValueWidget.create( plug["mode"] ) )
+		self.__row.append( GafferUI.PlugValueWidget.create( plug["value"] ), expand = True )
+
+		self._updateFromPlug()
+
+	def setPlug( self, plug ) :
+
+		GafferUI.PlugValueWidget.setPlug( self, plug )
+
+		self.__row[0].setPlug( plug["name"] )
+		self.__row[1].setPlug( plug["enabled"] )
+		self.__row[2].setPlug( plug["mode"] )
+		self.__row[3].setPlug( plug["value"] )
+
+	def hasLabel( self ) :
+
+		return True
+
+	def childPlugValueWidget( self, childPlug, lazy=True ) :
+
+		for w in self.__row :
+			if w.getPlug().isSame( childPlug ) :
+				return w
+
+		return None
+
+	def setReadOnly( self, readOnly ) :
+
+		if readOnly == self.getReadOnly() :
+			return
+
+		GafferUI.PlugValueWidget.setReadOnly( self, readOnly )
+
+		for w in self.__row :
+			w.setReadOnly( readOnly )
+
+	def _updateFromPlug( self ) :
+
+		with self.getContext() :
+			enabled = self.getPlug()["enabled"].getValue()
+
+		for i in ( 0, 2, 3 ) :
+			self.__row[i].setEnabled( enabled )
+
+	@staticmethod
```

Comment:
Hmm, I definitely don't feel strongly about the API here, but I was trying to keep the existing UI for LightTweaks, where all the custom parameters are listed at the root level of the menu.  I think it's important for the camera tweaks that the custom parameter options be tucked away inside a "/Custom Parameters" submenu - and rearranging these standard menu options can't really be done from tweaksRequestSignal?

Should I just duplicate the code for the list of custom parameters into LightTweaks and CameraTweaks, rather than having them in defaultTypesMenu?

Or is it OK to have defaultTypesMenu, if it just returns a list of tuples, and LightTweaks and CameraTweaks fetch it and then return it through tweaksRequestSignal()?

Reply:
> I think it's important for the camera tweaks that the custom parameter options be tucked away inside a "/Custom Parameters" submenu - and rearranging these standard menu options can't really be done from tweaksRequestSignal?

Couple of options :

- Always put them in a submenu if the signal produces some dedicated tweaks. This changes the LightTweaks menu, but I don't think it'll be any the worst for it.
- We use a metadata entry to provide an optional prefix for the custom parameters submenu.

> Should I just duplicate the code for the list of custom parameters into LightTweaks and CameraTweaks, rather than having them in defaultTypesMenu?

I think this is what we had originally agreed. There are so many other considerations in the camera work that deferring unnecessary decisions till later (when we also have other clients like ShaderTweaks to inform the decision) seemed a good strategy to me. The options I listed above don't seem terrible though - let's chat later...